lib/matobjnz.gi

```diff
@@ -0,0 +1,1422 @@
+  function( v, p, ob )
+    v![ELSPOS][p] := Int(ob);
+  end );
+
+InstallMethod( \{\}, "for a zmodnz vector and a list",
+  [ IsZmodnZVectorRep, IsList ],
+  function( v, l )
+    return Objectify(TypeObj(v),[v![BDPOS],v![ELSPOS]{l}]);
+  end );
+
+InstallMethod( PositionNonZero, "for a zmodnz vector", [ IsZmodnZVectorRep ],
+  function( v )
+    return PositionNonZero( v![ELSPOS] );
+  end );
+
+InstallMethod( PositionNonZero, "for a zmodnz vector", [ IsZmodnZVectorRep ],
+  function( v )
+    return PositionNonZero( v![ELSPOS] );
+  end );
+
+InstallOtherMethod( PositionNonZero, "for a zmodnz vector and start",
+  [ IsZmodnZVectorRep,IsInt ],
+  function( v,s )
+    return PositionNonZero( v![ELSPOS],s );
+  end );
+
+InstallMethod( PositionLastNonZero, "for a zmodnz vector", [ IsZmodnZVectorRep ],
+  function( v )
+    local els,i;
+    els := v![ELSPOS];
+    i := Length(els);
+    while i > 0 and IsZero(els[i]) do i := i - 1; od;
+    return i;
+  end );
+
+InstallMethod( ListOp, "for a zmodnz vector", [ IsZmodnZVectorRep ],
+function( v )
+local fam;
+  fam:=ElementsFamily(FamilyObj(v));
+  return List([1..Length(v![ELSPOS])],x->ZmodnZObj(fam,v![ELSPOS][x]));
+end );
+
+InstallMethod( ListOp, "for a zmodnz vector and a function",
+  [ IsZmodnZVectorRep, IsFunction ],
+function( v, f )
+local fam;
+  fam:=ElementsFamily(FamilyObj(v));
+  return List(List([1..Length(v![ELSPOS])],x->ZmodnZObj(fam,v![ELSPOS][x])),f);
+end );
+
+InstallMethod( Unpack, "for a zmodnz vector",
+  [ IsZmodnZVectorRep ],
+function( v )
+local fam;
+  fam:=ElementsFamily(FamilyObj(v));
+  return List([1..Length(v![ELSPOS])],x->ZmodnZObj(fam,v![ELSPOS][x]));
+end );
+
+############################################################################
+# Arithmetical operations:
+############################################################################
+
+InstallMethod( \+, "for two zmodnz vectors",
+  [ IsZmodnZVectorRep, IsZmodnZVectorRep ],
+function( a, b )
+local ty,i,m,mu;
+  if not IsMutable(a) and IsMutable(b) then
+      ty := TypeObj(b);
+  else
+      ty := TypeObj(a);
+  fi;
+  m:=Size(a![BDPOS]);
+  b:=SUM_LIST_LIST_DEFAULT(a![ELSPOS],b![ELSPOS]);
+  if not IsMutable(b) then mu:=true;b:=ShallowCopy(b);
+  else mu:=false;fi;
+  for i in [1..Length(b)] do if b[i]>=m then b[i]:=b[i] mod m;fi;od;
+  if mu then MakeImmutable(b);fi;
+  return Objectify(ty,[a![BDPOS],b]);
+end );
+
+InstallMethod( \-, "for two zmodnz vectors",
+  [ IsZmodnZVectorRep, IsZmodnZVectorRep ],
+function( a, b )
+local ty,i,m;
+  if not IsMutable(a) and IsMutable(b) then
+      ty := TypeObj(b);
+  else
+      ty := TypeObj(a);
+  fi;
+  m:=Size(a![BDPOS]);
+  b:=a![ELSPOS] - b![ELSPOS];
+  for i in [1..Length(b)] do if b[i]<0 then b[i]:=b[i] mod m;fi;od;
+  return Objectify(ty,[a![BDPOS],b]);
+end );
+
+InstallMethod( \=, "for two zmodnz vectors",
+  [ IsZmodnZVectorRep, IsZmodnZVectorRep ],
+  function( a, b )
+    return EQ_LIST_LIST_DEFAULT(a![ELSPOS],b![ELSPOS]);
```

Comment:
Is it intended that two VectorObj over different rings are regarded as equal (and can be added, etc.)?

Reply:
AH -- did not spot that a test for the families is missing -- will add.